### PR TITLE
Add ContributersSection #47

### LINKED ISSUE #47: Add ContributersSection component
Open source repositories like this are only possible due to contributions from people like you. Let's give a shout-out and add everyone to the README file. And let's do it automatically, so that new contributors in the future will also be added automatically.

## Describe the solution you'd like

Update the `ContributersSection` component so that it can still be included as `<ContributorsSection />` and renders a full section similar to the [ContributingSection](https://github.com/dbartholomae/jsx-readme/blob/main/src/components/ContributingSection.tsx), including heading, emoji and line breaks at the end. It should display the avatars of the 12 newest contributors, each linking to their GitHub profile and above it the following sentence:

```md
This package only works thanks to [all of our contributors](https://github.com/dbartholomae/jsx-readme/graphs/contributors).
```

After the avatars, if there are more than 12 contributors, e. g. 1514 in total, it should show
```md
[+ 1,502 contributors](https://github.com/dbartholomae/jsx-readme/graphs/contributors)
```
with `1,502` being `1514 - 12`.

It should take an optional prop for a Github access token in case it needs one, but should try to read it from the default environmental variables if the prop is not set. In case we don't even need a token (assuming a public repository), then feel free to skip adding the prop.

The `<ContributorsSection />` should be included in the [Readme example script](https://github.com/dbartholomae/jsx-readme/blob/main/examples/README.md.tsx) right above the `<ContributingSection />`.

You might have to also edit the [build workflow](https://github.com/dbartholomae/jsx-readme/blob/main/.github/workflows/build.yml) to use `env: GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}` for the `Build documentation` step so that, when the Readme is built, the script has access to the token.

## Describe alternatives you've considered

I searched for contributor overview pages we could just link to or images we could just embed, but didn't find any nice ones.

## Additional context

This will require the use of the new `jsx-md` async API via `awaitComponent` which can be seen in the [ExamplesFromPkg component](https://github.com/dbartholomae/jsx-readme/blob/main/src/components/ExamplesFromPkg.tsx).

You will also need to familiarize yourself with the collaborators from the [GitHub GraphQL API](https://docs.github.com/en/free-pro-team@latest/graphql/reference/objects#repository) or the [outdated GitHub Rest API](https://developer.github.com/v3/repos/#list-repository-contributors).

### PR DESCRIPTION
**Pull request by junwen-k**
Closes #47

## Checklist:
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.

Hi :)
There are a few issues that I've encountered when doing this.
For a pretty contributors section, we might need to use raw HTML element, which current `jsx-md` does not support (As far as I know)

The GitHub build workflow might not be correct as well.

### CODE REVIEW

File: src/containers/ContributorsSectionFromPkg.tsx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  pkg: PackageJSON;
+  githubAccessToken?: string;
+  contributorsMaxDisplayLength?: number;
+}
+/** Displays a contributors section from a package.json file. */
+export const ContributorsSectionFromPkg: Component<Props> = awaitComponent(
+  async ({
+    title = "‍💼 Contributors",
+    pkg,
+    githubAccessToken,
```
Comment (dbartholomae) [inline src/containers/ContributorsSectionFromPkg.tsx:79]:
```suggestion
    githubAccessToken = process.env.GITHUB_TOKEN,
```

File: src/containers/ContributorsSectionFromPkg.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    contributorsMaxDisplayLength = 12,
+  }) => {
+    const repoData = extractGithubOwnerAndRepo(pkg.repository);
+    if (repoData === undefined) {
+      return null;
+    }
+
+    const { repo, owner } = repoData;
+
+    const auth = githubAccessToken || process.env.GITHUB_TOKEN;
```
Comment (dbartholomae) [inline src/containers/ContributorsSectionFromPkg.tsx:89]:
```suggestion
```

File: src/containers/ContributorsSectionFromPkg.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    const octokit = new Octokit({ auth });
```
Comment (dbartholomae) [inline src/containers/ContributorsSectionFromPkg.tsx:91]:
```suggestion
    const octokit = new Octokit({ auth: githubAccessToken });
```

File: src/containers/ContributorsSectionFromPkg.tsx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    if (resp.data.length <= 0) {
+      return contributors;
+    }
+    contributors = contributors.concat(
+      resp.data
+        // NOTE: Exclude "Bot" and repo owner as contributors does'nt work as total contributors count will include them as well.
+        // .filter(
+        //   (contributor) =>
+        //     contributor.type !== "Bot" && contributor.login !== owner
+        // )
```
Comment (dbartholomae) [inline src/containers/ContributorsSectionFromPkg.tsx:56]: Not sure why we can't filter out bots:

```suggestion.filter((contributor) => contributor.type !== "Bot")
```

This way, bots should now be in the `additionalContributorsCount`, if I understood the code correctly

File: test/README.expected.md [hunk 5]
```diff
@@ -232,6 +232,21 @@ If you are interested in contributing to this repository, please read up on the
...
+[![MDShields7](https://avatars.githubusercontent.com/u/32501364?size=100 "MDShields7")](https://github.com/MDShields7 "MDShields7's GitHub profile'")
+[![AJMcDee](https://avatars.githubusercontent.com/u/22745495?size=100 "AJMcDee")](https://github.com/AJMcDee "AJMcDee's GitHub profile'")
+[![tancredosouza](https://avatars.githubusercontent.com/u/13404051?size=100 "tancredosouza")](https://github.com/tancredosouza "tancredosouza's GitHub profile'")
+[![Ruchika30](https://avatars.githubusercontent.com/u/16018506?size=100 "Ruchika30")](https://github.com/Ruchika30 "Ruchika30's GitHub profile'")
+[![anshdhinhgra47](https://avatars.githubusercontent.com/u/56676059?size=100 "anshdhinhgra47")](https://github.com/anshdhinhgra47 "anshdhinhgra47's GitHub profile'")
+[![gedaiu](https://avatars.githubusercontent.com/u/4217037?size=100 "gedaiu")](https://github.com/gedaiu "gedaiu's GitHub profile'")
+[![i-am-jeetu](https://avatars.githubusercontent.com/u/66921514?size=100 "i-am-jeetu")](https://github.com/i-am-jeetu "i-am-jeetu's GitHub profile'")
+
+[+ 2 contributors](https://github.com/dbartholomae/jsx-readme/graphs/contributors)
+
```
Comment (dbartholomae) [inline test/README.expected.md:249]: What can we do to make this part of the test less brittle?

File: src/components/ContributorsSection.test.tsx [hunk 6]
```diff
@@ -5,35 +5,74 @@ import { ContributorsSection } from "./ContributorsSection";
...
+      await render(
+        <ContributorsSection
+          contributors={[]}
+          owner="dbartholomae"
+          repo="test-repo"
+          additionalContributorsCount={1}
+        />
+      )
+    ).toContain(
+      "[+ 1 contributors](https://github.com/dbartholomae/test-repo/graphs/contributors)"
```
Comment (dbartholomae) [inline src/components/ContributorsSection.test.tsx:75]: Please add a test for what the main section looks like with an empty contributors array. In that case, it could make sense to include the number of contributors directly in the first sentence instead of doubling the link.
I know it's an edge case, but it should be easy enough to fix :)

File: src/components/ContributorsSection.test.tsx [hunk 7]
```diff
@@ -5,35 +5,74 @@ import { ContributorsSection } from "./ContributorsSection";
...
+    expect(
+      await render(
+        <ContributorsSection
+          contributors={contributors}
+          owner="dbartholomae"
+          repo="test-repo"
+        />
+      )
+    ).toContain(
+      '[![dbartholomae](https://avatars.githubusercontent.com/u/1?size=100 "dbartholomae")](https://github.com/dbartholomae "dbartholomae\'s GitHub profile\'")'
```
Comment (dbartholomae) [inline src/components/ContributorsSection.test.tsx:60]: The sizing of the avatars doesn't seem to work correctly as you can see when you scroll to the README of the branch:
https://github.com/junwen-k/jsx-readme/tree/feat/%2347

**Review by dbartholomae — CHANGES_REQUESTED** [review]
Thanks, great work! There's five more things to do:
1. Fix the sizing of the avatars
2. Find a way to make the integration test less brittle
3. Exclude bots
4. Explicitly handle the case where contributors is an empty array
5. Small refactor

File: src/components/ContributorsSection.test.tsx (hunk 7 above)
Comment (junwen-k) [reply src/components/ContributorsSection.test.tsx:60]: Do you have any recommendation to workaround this issue? I've did some research on other contributor list and they usually involve some HTML code. As far as I know we are using `jsx-md` and it only outputs markdown without any raw HTML.

Ideally, my thinking is that the component should dynamically spans how many avatars in one row based on the number of contributors, but using only Markdown it seems like a challenge. I've tried using `<Table />` component from `jsx-md` and it seems like doesn't work very well in this case.

The way I try to hack it now is by explicitly setting the size of the avatar using GitHub avatar server, but apparently it does not work if there's no custom avatar image provided. See comments on https://stackoverflow.com/a/36380674.

See [all-contributors](https://github.com/all-contributors/all-contributors) and [markdown-contributors](https://thodorisbais.github.io/markdown-contributors/) for example.

Comment (dbartholomae) [reply src/components/ContributorsSection.test.tsx:60]: Hi, it's fine to use html and create a `HtmlImg` component in this library. I might move it to the other library at some point, but for now let's not limit the functionality of `jsx-readme` by only using components from `jsx-md`. In the end, components can just return text.

Comment (junwen-k) [reply src/components/ContributorsSection.test.tsx:60]: I see, I will try to implement that :) Not sure if I know how to do it or not though. Thank you!

File: test/README.expected.md (hunk 5 above)
Comment (junwen-k) [reply test/README.expected.md:249]: One way is that instead of using `<ContributorsSectionFromPkg />` component, we can revert back to the previous `<ContributorsSection />` component which does not rely on dynamic content from GitHub API. But that defeats the purpose of having that component and I am not sure if there's other workaround though...

Comment (dbartholomae) [reply test/README.expected.md:249]: I'll added a test-fixture repo to GitHub: https://github.com/dbartholomae/jsx-readme-test-fixture
If you could create a PR (I will mark it manually as hacktoberfest-approved) with a commit to that repo as well (just a link back to this repo in the readme should do, but at least two contributors might be helpful for the test), then change the [`package.json`](https://github.com/dbartholomae/jsx-readme/tree/main/examples) to use the test fixture repo, this should solve it.

File: src/components/ContributorsSection.test.tsx (hunk 7 above)
Comment (dbartholomae) [reply src/components/ContributorsSection.test.tsx:60]: Something along these lines:
```tsx
// in src/elements/HtmlImage.tsx

/* @jsx MD */
import MD, { Component, Fragment } from "..";

/** @internal */
interface Props {
  children: string;
  src: string;
}

/**
 * @example
 *   ```js
 *   render(<HtmlImage src="./image.png" title="Title">
 *     Alternative text
 *   </Image>)
 *   ===
 *   '<img alt="Alternative text" src="./image.png">'
 */
export const HtmlImage: Component<Props> = ({ children, src }) => (
  <Fragment>
    {'<img alt="'}
    {children}
    {'" src="'}
    {src}
    {'">'}
  </Fragment>
);
```
(based on the [jsx-md image component](https://github.com/dbartholomae/jsx-md/blob/main/src/elements/Image.tsx), you can also add a [test](https://github.com/dbartholomae/jsx-md/blob/main/src/elements/Image.test.tsx) to confirm that it works)

Comment (dbartholomae) [reply src/components/ContributorsSection.test.tsx:60]: Or, if you need more complicated things with tables etc. (similar to all-contributors), you can add a `Contributor` component and add whatever HTML you need.

File: src/containers/ContributorsSectionFromPkg.tsx (hunk 4 above)
Comment (junwen-k) [reply src/containers/ContributorsSectionFromPkg.tsx:56]: Yes you are right, previous I thought that the total contributors displaying does not match the `contributorsMaxDisplayLength` prop provided. Looks like it is working as expected because Bots and owner will be added into the additional contributors count instead.

**Comment by junwen-k** [comment]
I've updated my PR and you might want to take a look at the README.md generation test update. Since we are switching to use `jsx-readme-test-fixture` repo, we cannot rely on that test README.md to be this repo's actual README file, hence I've separated the test and the actual README generation script.

As a side note, https://github.com/dbartholomae/jsx-readme-test-fixture is not currently participating `hacktoberfest` and you might want to add the topic. It will also be very helpful if you could flag my PR with the `hacktoberfest-accepted` label :)

File: README.md [hunk 8]
```diff
@@ -125,7 +125,7 @@ Create a README.MD template (you may copy the example from this repo examples/RE
 ```tsx
 // We need to tell the JSX transpiler that in this file,
 // instead of React we use the custom createElement and Fragment
-// functions from jsx-readme
+// functions from jsx-readme-test-fixture
```
Comment (dbartholomae) [inline README.md:128]:
```suggestion
// functions from jsx-readme
```

File: README.md [hunk 9]
```diff
@@ -173,10 +173,10 @@ const Readme: Component = () => (
     <DescriptionFromPkg pkg={pkg} />
     {/* You can use the components from jsx-md to build custom markdown. */}
     <Heading level={2}>🛠 Installation</Heading>
-    Add <InlineCode>jsx-readme</InlineCode> and <InlineCode>ts-node</InlineCode>{" "}
+    Add <InlineCode>jsx-readme-test-fixture</InlineCode> and <InlineCode>ts-node</InlineCode>{" "}
```
Comment (dbartholomae) [inline README.md:176]:
```suggestion
    Add <InlineCode>jsx-readme</InlineCode> and <InlineCode>ts-node</InlineCode>{" "}
```

File: README.md [hunk 10]
```diff
@@ -173,10 +173,10 @@ const Readme: Component = () => (
...
     to your <InlineCode>devDependencies</InlineCode>.
     <LineBreak />
-    <CodeBlock language="sh">npm i jsx-readme ts-node -D</CodeBlock>
+    <CodeBlock language="sh">npm i jsx-readme-test-fixture ts-node -D</CodeBlock>
```
Comment (dbartholomae) [inline README.md:179]:
```suggestion
    <CodeBlock language="sh">npm i jsx-readme ts-node -D</CodeBlock>
```

File: examples/README.md.tsx [hunk 11]
```diff
@@ -1,6 +1,6 @@
 // We need to tell the JSX transpiler that in this file,
 // instead of React we use the custom createElement and Fragment
-// functions from jsx-readme
+// functions from jsx-readme-test-fixture
```
Comment (dbartholomae) [inline examples/README.md.tsx:3]:
```suggestion
// functions from jsx-readme
```

File: examples/README.md.tsx [hunk 12]
```diff
@@ -48,10 +48,10 @@ const Readme: Component = () => (
     <DescriptionFromPkg pkg={pkg} />
     {/* You can use the components from jsx-md to build custom markdown. */}
     <Heading level={2}>🛠 Installation</Heading>
-    Add <InlineCode>jsx-readme</InlineCode> and <InlineCode>ts-node</InlineCode>{" "}
+    Add <InlineCode>jsx-readme-test-fixture</InlineCode> and <InlineCode>ts-node</InlineCode>{" "}
     to your <InlineCode>devDependencies</InlineCode>.
     <LineBreak />
-    <CodeBlock language="sh">npm i jsx-readme ts-node -D</CodeBlock>
+    <CodeBlock language="sh">npm i jsx-readme-test-fixture ts-node -D</CodeBlock>
```
Comment (dbartholomae) [inline examples/README.md.tsx:54]:
```suggestion
    <CodeBlock language="sh">npm i jsx-readme ts-node -D</CodeBlock>
```

File: examples/package.json [hunk 13]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "jsx-readme",
+  "name": "jsx-readme-test-fixture",
```
Comment (dbartholomae) [inline examples/package.json:2]:
```suggestion
  "name": "jsx-readme",
```

File: README.md [hunk 14]
```diff
@@ -54,7 +54,7 @@ Create a README.MD template (you may copy the example from this repo examples/RE
 
 ```json
 {
-  "name": "jsx-readme",
+  "name": "jsx-readme-test-fixture",
```
Comment (dbartholomae) [inline README.md:57]:
```suggestion
  "name": "jsx-readme",
```

File: test/README.expected.md [hunk 15]
```diff
@@ -54,7 +54,7 @@ Create a README.MD template (you may copy the example from this repo examples/RE
 
 ```json
 {
-  "name": "jsx-readme",
+  "name": "jsx-readme-test-fixture",
```
Comment (dbartholomae) [inline test/README.expected.md:57]:
```suggestion
  "name": "jsx-readme",
```

File: test/README.expected.md [hunk 16]
```diff
@@ -125,10 +125,10 @@ Create a README.MD template (you may copy the example from this repo examples/RE
 ```tsx
 // We need to tell the JSX transpiler that in this file,
 // instead of React we use the custom createElement and Fragment
-// functions from jsx-readme
+// functions from jsx-readme-test-fixture
```
Comment (dbartholomae) [inline test/README.expected.md:128]:
```suggestion
// functions from jsx-readme
```

File: test/README.expected.md [hunk 17]
```diff
@@ -173,10 +173,10 @@ const Readme: Component = () => (
     <DescriptionFromPkg pkg={pkg} />
     {/* You can use the components from jsx-md to build custom markdown. */}
     <Heading level={2}>🛠 Installation</Heading>
-    Add <InlineCode>jsx-readme</InlineCode> and <InlineCode>ts-node</InlineCode>{" "}
+    Add <InlineCode>jsx-readme-test-fixture</InlineCode> and <InlineCode>ts-node</InlineCode>{" "}
```
Comment (dbartholomae) [inline test/README.expected.md:176]:
```suggestion
    Add <InlineCode>jsx-readme</InlineCode> and <InlineCode>ts-node</InlineCode>{" "}
```

File: test/README.md.test.tsx [hunk 18]
```diff
@@ -2,14 +2,18 @@ import { promises as fs } from "fs";
 import { runTypeScriptFile } from "./utils/runTypeScriptFile";
 
 describe("README creation script", () => {
-  it("creates the README file of the repo", async () => {
+  afterAll(async () => {
+    await fs.unlink("./test/README.actual.md");
+  }, 20000);
```
Comment (dbartholomae) [inline test/README.md.test.tsx:7]: Do we need this? Having the file around in case the test fails was quite useful in the past. If need something like this, I would prefer to have it in a `beforeEach`

File: README.md [hunk 19]
```diff
@@ -232,6 +232,10 @@ If you are interested in contributing to this repository, please read up on the
...
...e="gedaiu's GitHub profile'" href="https://github.com/gedaiu"><img alt="gedaiu" src="https://avatars.githubusercontent.com/u/4217037?s=100" width="100px"></a><br /><a title="gedaiu's GitHub profile'" href="https://github.com/gedaiu"><sub><b>gedaiu</b></sub></a></td><td align="center"><a title="i-am-jeetu's GitHub profile'" href="https://github.com/i-am-jeetu"><img alt="i-am-jeetu" src="https://avatars.githubusercontent.com/u/66921514?s=100" width="100px"></a><br /><a title="i-am-jeetu's GitHub profile'" href="https://github.com/i-am-jeetu"><sub><b>i-am-jeetu</b></sub></a></td><td align="center"><a title="Pipo93's GitHub profile'" href="https://github.com/Pipo93"><img alt="Pipo93" src="https://avatars.githubusercontent.com/u/12968697?s=100" width="100px"></a><br /><a title="Pipo93's GitHub profile'" href="https://github.com/Pipo93"><sub><b>Pipo93</b></sub></a></td><td align="center"><a title="jruipinto's GitHub profile'" href="https://github.com/jruipinto"><img alt="jruipinto" src="https://avatars.githubusercontent.com/u/44852805?s=100" width="100px"></a><br /><a title="jruipinto's GitHub profile'" href="https://github.com/jruipinto"><sub><b>jruipinto</b></sub></a></td></tr></table>
```
Comment (dbartholomae) [inline README.md:235]: Looks really nice! It seems that semantic-release-bot isn't filtered out, though :-/

File: src/containers/ContributorsSectionFromPkg.tsx [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+    );
+    // No more contributors.
+    if (resp.data.length <= 0) {
+      return contributors;
+    }
+    contributors = contributors.concat(
+      resp.data
+        .filter(
+          (contributor) =>
+            contributor.type !== "Bot" && contributor.login !== owner
```
Comment (dbartholomae) [inline src/containers/ContributorsSectionFromPkg.tsx:54]: From the examples I've seen so far it seems that the owner is usually also considered a contributor.
```suggestion
            contributor.type !== "Bot"
```

**Review by dbartholomae — CHANGES_REQUESTED** [review]
Thanks, looks really good! I've got a couple of small changes, and two questions left.

File: jest.config.js [hunk 21]
```diff
@@ -16,4 +16,5 @@ module.exports = {
   },
   testRegex: ".*test\\.(t|j)sx?$",
   moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json", "node"],
+  modulePathIgnorePatterns: ["src/containers/"],
```
[the remark this answers is not in the document]
Comment (junwen-k) [reply jest.config.js]: Haha, this was added before test-fixture repo was added. Will add test based on that repo.

File: test/README.md.test.tsx (hunk 18 above)
Comment (junwen-k) [reply test/README.md.test.tsx]: Removed test cleanup

File: jest.config.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (junwen-k) [reply jest.config.js]: I have one last problem, 

```
---------------------------------|---------|----------|---------|---------|-------------------
File                             | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
---------------------------------|---------|----------|---------|---------|-------------------
 src/containers                  |   91.66 |    78.94 |     100 |   91.42 |                   
  ContributorsSectionFromPkg.tsx |   91.66 |    78.94 |     100 |   91.42 | 24,28,60
```

Apparently ContributorsSectionFromPkg could not achieve 100% test coverage. Based on the uncovered line(s) stated, those line were 
```ts
  if (!resp.headers.link) {
    return 0;
  }
```
```ts
  const links = parseLinkHeader(resp.headers.link);
  if (!links) {
    return 0;
  }
```
and 
```ts
  return contributors.slice(0, fetchUntil);
```
Do you have any ideas how to achieve 100% test coverage? Or should I just add the istanbul ignore line comment.

[the remark this answers is not in the document]
Comment (dbartholomae) [reply jest.config.js]: Yes, ignore-comments with a short explanation of why this is hard to test should be enough.

**Comment by junwen-k** [comment]
Updated PR, all test runs successfully now :)
